bpf/lib/nat.h

```diff
@@ -295,8 +310,30 @@ snat_v4_nat_handle_mapping(struct __ctx_buff *ctx,
 	}
 
 	if (*state) {
-		barrier_data(*state);
-		return 0;
+		int ret;
+		struct ipv4_ct_tuple rtuple = {};
+
+		if (target->addr == (*state)->to_saddr &&
+		    needs_ct == (*state)->common.needs_ct) {
+			barrier_data(*state);
+			return 0;
+		}
+
+		/* Recreate the SNAT and RevSNAT entries if the source IP is stale.
+		 * Otherwise, the packet will be erroneously SNATed with the stale
+		 * source IP.
+		 */
+		ret = __snat_delete(map, tuple);
+		if (IS_ERR(ret))
+			return ret;
```

Comment:
I think we may have a race condition here. When two or more threads execute this code path in parallel, one may hit `ENOENT` here and returns with error. Is this what suppose to happen? The `ENOENT` may not an error here, because we want to delete the stale entry, but it was accomplished by someone else, which sounds fine.

Reply:
If  the preceding thread already remove the stale entry and  `__snat_delete` returns `ENOENT`, we might be able to lookup SNAT entry again and proceed. However we can't  guarantee that the earlier thread finishes recreating the SNAT entry at this point.

It seems like multiple threads hitting the same stale tuple with a different desired src IP simultaneously is a rare situation, so it might be acceptable to error out in such cases, IMHO.

Reply:
Sure, agree that it's a rare situation (because of RSS). I don't have a strong opinion around this.